### ISSUE TITLE
Automatic paket restore is slow while building solution from Visual Studio.

### ISSUE THREAD

**Issue by alexanium-dev** [fsprojects/Paket#3616]
### Description

Visual Studio restores packages for every project individually during build. This leads to a significant build slowdown. The build output looks like following (all packages are already restored):

> 101>  Paket version 5.215.0
> 101>  <solution-path>\paket-files\paket.processlock is locked by paket.exe (PID = 13380). Waiting...
> 101>  Starting restore process.
> 101>  Performance:
> 101>   - Runtime: 2 seconds
> 101>  Lib101 -> <solution-path>\bin\Debug\Lib101.dll
> 100>  Paket version 5.215.0
> 100>  Starting restore process.
> 100>  Performance:
> 100>   - Runtime: 1 second

### Repro steps

Try to build solution in VisualStudio2017 with many projects (old style net472) and many referenced NuGet packages with auto-restore enabled. Then try to build it again.

### Expected behavior

Almost no slowdown if all packages already restored.

### Actual behavior

Second solution build with "auto-restore off" lasts a couple of seconds, with "auto-restore on" - a couple of minutes.

### Known workarounds

A. Disable auto-restore and restore manually if needed.
B. Upgrade to new project format
C. Use 5.210.1 Paket version

**Comment by matthid** [comment]
> with many projects (old style net472)

Isn't this "expected"? Afaik almost no work went into performance optimization for this scenario.
Another workaround is to upgrade your projects.

**Comment by newsash** [comment]
New project format isn't mature enough for a big solution with WPF and COM references. So I'll have to disable auto-restore and wait for a .NET Core 3.0 release.

I'm not good at msbuild stuff, but isn't it a error, that restore runs for every project every build instead of running once (per solution or per project) and then just comparing "PaketRestoreCachedHash" and "PaketRestoreLockFileHash"?

**Comment by matthid** [comment]
Yes that optimization could probably be implemented. It's just that our integration into old style project looks quite different and uses it's own targets file.
As there were some changes recently it is possible something broke. Did this work with an older version?

**Comment by newsash** [comment]
I suppose, the problem is in the following lines:
```
    <PropertyGroup Condition="Exists('$(PaketRestoreCacheFile)') ">
      <PaketRestoreCachedHash>$([System.IO.File]::ReadAllText('$(PaketRestoreCacheFile)'))</PaketRestoreCachedHash>
      <PaketRestoreLockFileHash>$([System.IO.File]::ReadAllText('$(PaketLockFilePath)'))</PaketRestoreLockFileHash>
      <PaketRestoreRequired>true</PaketRestoreRequired>
      <PaketRestoreRequired Condition=" '$(PaketRestoreLockFileHash)' == '$(PaketRestoreCachedHash)' ">false</PaketRestoreRequired>
      <PaketRestoreRequired Condition=" '$(PaketRestoreLockFileHash)' == '' ">true</PaketRestoreRequired>
    </PropertyGroup>
```

It compares `paket.lock` with the following text (and result is always false):
```
{
  "packagesDownloadedHash": "2F1DF8C8CE894EDA4B3A081969F918FE1F860F393E1F3B8A221E62AB08CF846A",
  "projectsRestoredHash": "2F1DF8C8CE894EDA4B3A081969F918FE1F860F393E1F3B8A221E62AB08CF846A"
}
```

**Comment by matthid** [comment]
Indeed, it needs the same "update" as the other targets file. Can you send a PR? Just a word of warning: The update only works with recent msbuild versions (I think msbuild >= 15.8), so if you still use an older msbuild version the fix will not change anything for you.

**Comment by newsash** [comment]
I don't know if I'll find a time to make a good PR (I think it is possible to make a polyfill for GetFileHash).

Another workaround is to use 5.210.1 version.

**Comment by matthid** [comment]
> I think it is possible to make a polyfill for GetFileHash

How?

**Comment by newsash** [comment]
> How?

It's just an idea, I don't know if it'll work. Something like this:
```
<UsingTask
  TaskName="GetFileHash"
  TaskFactory="CodeTaskFactory"
  AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll" 
  Condition=" '$(MSBuildVersion)' &lt; '15.8' ">
  <ParameterGroup>
    <Algorithm ParameterType="System.String" Required="False" Output="False"/>
    <Items ParameterType="Microsoft.Build.Framework.ITaskItem[]" Required="True" Output="False"/>
    ... other parameters from https://github.com/microsoft/msbuild/blob/master/ref/Microsoft.Build.Tasks.Core/netstandard/Microsoft.Build.Tasks.Core.cs#L370
  </ParameterGroup>
  <Task>
    // Polyfill implementation
    <Reference Include="System.Xml"/>
    <Using Namespace="System"/>
    <Using Namespace="System.IO"/>
    <Code Type="Fragment" Language="cs">
      <![CDATA[
      // Display "Hello, world!"
      Log.LogError("Hello, world!");
      ]]>
    </Code>
  </Task>
</UsingTask>
```

Inline tasks documentation: https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-inline-tasks?view=vs-2019
Example from 2012: http://blog.ylett.com/2012/05/msbuild-task-to-generate-hash-of-files.html (another AssemblyFile, same API).

**Comment by xt1** [comment]
`certutil -hashfile ..\paket.lock sha1 | find /V "hash"` will dump the SHA1 hash on windows boxen.

ref: https://superuser.com/questions/245775/is-there-a-built-in-checksum-utility-on-windows-7

**Comment by newsash** [comment]
Decided to use 5.210.1 until upgrading projects. Should I close this issue?

**Comment by xt1** [comment]
I think it's worth keeping open, but maybe rename the issue: paket auto restore restores for every project in solution every time  

The root issue here is the hashing. Instead of hashing, paket should just copy the lock file to paket-files/paket.lock.restored.cached.  This would make it easy for MSBuild to compare the files and determine if the restore needs to run again.

I'm assuming that we want to avoid calling paket.exe to figure out if we need to run restore, since every process exec costs a little bit.
